stdlib/sys.mli

```diff
@@ -421,6 +423,10 @@ external opaque_identity : 'a -> 'a = "%opaque"
       done
     ]}
 
+    Another use is to prevent FFI-shared data from being garbage collected
+    before a certain point by forcing liveness analysis to consider the value
+    alive until the call completes.
```

Comment:
I find that this misses some context. In order to use data from the FFI, you generally need to register the value as a root from the FFI. This is needed not only to keep the value alive but also to update the value in case it is moved by the GC. So it is incorrect to advertise opaque_identity as a general solution to this problem. Are you specifically talking about the situation where some data:
- is allocated on the foreign heap
- is referenced from an OCaml value (e.g. Custom block)
- is deallocated from a finaliser attached to this OCaml value

?

Reply:
I apologize: for context, see https://github.com/yallop/ocaml-ctypes/issues/571 but I was using the term "FFI" here really as shorthand for using Ctypes.

I believe the data is allocated on the foreign heap _from ocaml_, then stored in an ocaml value (unsure if it's a custom block), and then yes, deallocated from a finaliser.

The situation I'm talking about specifically is, I believe what you describe in your bullet-points, but perhaps @yallop  can add more context here that I'm might be missing.

Reply:
When writing C bindings by hand the liveness of OCaml values is a bit more obvious (e.g. the C stubs themselves register the arguments as roots with the `CAMLparam` macro).

However when using Ctypes one should be careful about the lifetime of OCaml values that are used by functions called through Ctypes: having a value in scope is not enough to guarantee liveness, and the value passed to the FFI value may not be the value holding the memory allocated (e.g. it could be a pointer to an OCaml value that is a pointer to an array, but the actual array is held by another OCaml value). So a mechanism is needed to keep Ctypes allocated values alive across such FFI calls, otherwise the OCaml runtime may garbage collect values that are still in use by the FFI. (AFAIK there is no obvious equivalent to `CAMLparam` for Ctypes, other than manually managing GC roots)

`Sys.opaque_identity` would be a good choice *if* that guarantee is made explicit in the documentation.
One can then call `Sys.opaque_identity` with all the parameters that were passed to the FFI call (and all other OCaml values that actually "hold" the memory for those values).

As has been pointed out this guarantee wasn't true in older OCaml releases, but is true in current ones (e.g. that is why in this PR from 2018 extra code was needed https://github.com/xapi-project/xen-api/pull/3669/files, the compiler only got changed to guarantee liveness with `Sys.opaque_identity` in 2021).
If this behaviour is documented then we can be more confident in using it for this purpose, knowing that future OCaml versions will keep that guarantee.

Reply:
I have remove that paragraph.  I think the original paragraph I added above should be sufficient.

Reply:
I agree, mentioning the liveness guarantee should be enough for the purpose of Ctypes